File: src/debugLog.ts
/**
 * Full-fidelity debug trace — a pre-release observability aid.
 *
 * When `config.debugLog` is on (env `DAINTREE_ASSISTANT_DEBUG_LOG=1`), EVERYTHING
 * the assistant does is appended to a per-session human-readable file:
 *
 *     <logDir>/<YYYY-MM-DD>-<sessionId>.log   (logDir defaults to ~/.daintree/logs)
 *
 * That means every model request and response (full message arrays included),
 * every tool/function call with its arguments and result, and the whole watcher
 * lifecycle. These logs are intentionally large and complete — values are NOT
 * truncated — so you can reconstruct a session end to end.
 *
 * Each process gets its OWN dated+id file (no shared `debug.log`, no rotation): a
 * new instance never clobbers a previous run's log. {@link startDebugLog} opens the
 * file once at boot, writes a `session.start` header naming the project/tier/models,
 * and returns the path so the caller can announce "logging to <file>". As part of
 * boot it also deletes any log older than {@link MAX_LOG_AGE_MS} (7 days).
 *
 * The log directory is GLOBAL, not per-project, so one `tail -f` covers every
 * session regardless of which project it was bound to.
 *
 * Logging is a no-op when disabled, and it NEVER throws into the caller: a write
 * failure warns once on stderr and is otherwise swallowed, because losing a debug
 * line must never take down a model call, a tool dispatch, or a watcher check.
 */
import fs from "node:fs";
import path from "node:path";
import type { AppConfig } from "./config.js";

/** Logs older than this (by mtime) are deleted at boot. */
export const MAX_LOG_AGE_MS = 7 * 24 * 60 * 60 * 1000;

/** A field rendered inline (on the header line) must be a scalar this short. */
const INLINE_MAX = 120;

let warnedOnce = false;
/** The file the current process writes to, once logging has started. */
let activeLogPath: string | undefined;

/** Just the config fields the logger needs — keeps call sites and tests light. */
export type DebugLogConfig = Pick<AppConfig, "debugLog" | "logDir">;

function isoNow(): string {
  try {
    return new Date().toISOString();
  } catch {
    return String(Date.now());
  }
}

/** Session log filename: `<YYYY-MM-DD>-<id>.log`. */
function sessionLogFileName(id: string): string {
  const date = isoNow().slice(0, 10);
  const safeId = id.replace(/[^\w.-]/g, "") || "session";
  return `${date}-${safeId}.log`;
}

function randomId(): string {
  return Math.random().toString(36).slice(2, 10);
}

/**
 * The file `logDebug` should append to within `logDir`. Reuses the session file
 * opened by {@link startDebugLog}; if logging wrote before boot (e.g. a direct
 * `logDebug` in a test) or the active path belongs to a different `logDir`, it
 * lazily opens a fresh session file so writes still coalesce into one file.
 */
function resolveTarget(logDir: string): string {
  if (activeLogPath && path.dirname(activeLogPath) === logDir) return activeLogPath;
  activeLogPath = path.join(logDir, sessionLogFileName(randomId()));
  return activeLogPath;
}

/** The log file the current process is writing to, once logging has started. */
export function currentDebugLogPath(): string | undefined {
  return activeLogPath;
}

function safeStringify(v: unknown, indent?: number): string {
  try {
    return JSON.stringify(v, null, indent) ?? String(v);
  } catch {
    return String(v);
  }
}

/** A short, single-line scalar that can sit on the header line as `key=value`. */
function isInlineScalar(v: unknown): boolean {
  if (v === null || v === undefined) return true;
  if (typeof v === "number" || typeof v === "boolean") return true;
  if (typeof v === "string") return v.length <= INLINE_MAX && !v.includes("\n");
  return false;
}

/** Render a large/structured value as an indented block (full, untruncated). */
function blockValue(v: unknown): string {
  const s = typeof v === "string" ? v : safeStringify(v, 2);
  return s
    .split("\n")
    .map((l) => `    ${l}`)
    .join("\n");
}

function warnOnce(err: unknown): void {
  if (warnedOnce) return;
  warnedOnce = true;
  console.error(
    `[debugLog] write failed (logging disabled for this run): ${
      err instanceof Error ? err.message : String(err)
    }`,
  );
}

/**
 * Append one event to the current session log. No-op unless `cfg.debugLog` is on.
 * `event` is a short dotted name (e.g. "model.request", "tool.call",
 * "watcher.stop"); `fields` carries the structured payload. Short scalars render
 * inline on the header line; objects, arrays, and multi-line strings render as
 * indented blocks below it — full and untruncated.
 */
export function logDebug(
  cfg: DebugLogConfig | undefined,
  event: string,
  fields: Record<string, unknown> = {},
): void {
  // Tolerate a missing/partial config — logging must never break the caller.
  if (!cfg?.debugLog || !cfg.logDir) return;
  const ts = isoNow();
  try {
    fs.mkdirSync(cfg.logDir, { recursive: true });
    const target = resolveTarget(cfg.logDir);

    const inline: string[] = [];
    const blocks: string[] = [];
    for (const [k, v] of Object.entries(fields)) {
      if (v === undefined) continue; // omit absent fields — they're just noise
      if (isInlineScalar(v)) inline.push(`${k}=${v === null ? "null" : String(v)}`);
      else blocks.push(`  ${k}:\n${blockValue(v)}`);
    }

    let out = `${ts}  ${event}${inline.length ? `  ${inline.join(" ")}` : ""}\n`;
    if (blocks.length) out += `${blocks.join("\n")}\n`;
    fs.appendFileSync(target, out);
  } catch (err) {
    warnOnce(err);
  }
}

/**
 * Begin a debug-log session at assistant startup. No-op (returns undefined) when
 * logging is off. Otherwise:
 *   1. delete any log older than {@link MAX_LOG_AGE_MS} (7 days);
 *   2. open a fresh `<date>-<sessionId>.log`;
 *   3. write a `session.start` header naming the project/tier/models/MCP;
 * and return the file path so the caller can announce "logging to <path>". Call
 * once per process, after config is loaded.
 */
export function startDebugLog(cfg: AppConfig, sessionId?: string): string | undefined {
  if (!cfg.debugLog || !cfg.logDir) return undefined;
  pruneOldLogs(cfg.logDir);
  activeLogPath = path.join(cfg.logDir, sessionLogFileName(sessionId ?? randomId()));
  logDebug(cfg, "session.start", {
    project: cfg.projectPath,
    projectId: cfg.projectId,
    windowId: cfg.windowId,
    tier: cfg.tier,
    largeModel: cfg.largeModel,
    mediumModel: cfg.mediumModel,
    smallModel: cfg.smallModel,
    mcpUrl: cfg.mcpUrl ?? "(unset)",
    offline: cfg.offline,
    autoApprove: cfg.autoApprove,
    stateDir: cfg.stateDir,
    logDir: cfg.logDir,
    pid: process.pid,
    node: process.version,
  });
  return activeLogPath;
}

/** Matches our session log filenames (`<YYYY-MM-DD>-<id>.log`) so pruning never
 *  touches unrelated `*.log` files if logDir points at a shared directory. */
const SESSION_LOG_RE = /^\d{4}-\d{2}-\d{2}-.+\.log$/;

/** Delete our own session logs in `dir` older than {@link MAX_LOG_AGE_MS} by mtime.
 *  Never throws; only files matching {@link SESSION_LOG_RE} are eligible. */
function pruneOldLogs(dir: string): void {
  let entries: string[];
  try {
    entries = fs.readdirSync(dir);
  } catch {
    return; // directory doesn't exist yet — nothing to prune.
  }
  const cutoff = Date.now() - MAX_LOG_AGE_MS;
  for (const f of entries) {
    if (!SESSION_LOG_RE.test(f)) continue;
    const p = path.join(dir, f);
    try {
      if (fs.statSync(p).mtimeMs < cutoff) fs.unlinkSync(p);
    } catch {
      // Best-effort cleanup — a failure to stat/remove one file is harmless.
    }
  }
}

File: internal/debuglog/debuglog.go
// Package debuglog writes a per-session, append-only, human-readable trace. It is a
// no-op when disabled and NEVER panics into the caller: a write failure warns ONCE on
// stderr, then is swallowed.
//
// # Redaction
//
// Every value is passed through internal/redact before it is written, and every block
// value is capped. That happens HERE, at the write boundary, rather than at the ~30 call
// sites, for two reasons: a call site that forgets is invisible until the day it matters,
// and new call sites get the protection without knowing the rule exists.
//
// What redaction does and does not buy: credential SHAPES and this process's own
// registered secrets are removed, so a bearer token, an `export API_KEY=…`, or a PEM
// block will not survive into the file. Everything else does. The trace still contains
// the conversation, terminal output, file excerpts, issue bodies, and memory contents,
// which is exactly what makes it useful for archaeology and exactly why it is an
// owner-only local artifact (dir 0700, file 0600, pruned after 7 days) and NOT a support
// artifact. `daintree-assistant support-bundle` is the thing to hand to someone else.
package debuglog

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/daintreehq/assistant/internal/redact"
)

const (
	// maxLogAgeMs: logs older than this (by mtime) are deleted at boot.
	maxLogAgeMs = int64(7 * 24 * 60 * 60 * 1000)
	// inlineMax: a string ≤120 chars with no newline renders inline as key=value;
	// otherwise it renders as an indented block.
	inlineMax = 120
	// blockValueMax caps a single block value (a tool's args, a result payload, a
	// terminal excerpt). Generous — the whole point of this trace is that you can read
	// what the model actually saw, and a stingy cap would cut off the interesting part
	// of exactly the payload you opened the log for.
	//
	// But unbounded was worse. One terminal dump can be megabytes; a turn that reads
	// several of them wrote a log nobody could open, and the value that mattered was
	// buried under a screenful of build output. The overflow is replaced by a size and a
	// content hash, so two occurrences of the same payload are still recognisably the
	// same without either being stored.
	blockValueMax = 64 * 1024
)

// sessionLogRe matches the only filenames eligible for pruning.
var sessionLogRe = regexp.MustCompile(`^\d{4}-\d{2}-\d{2}-.+\.log$`)

// safeIDRe strips characters not allowed in a filename id.
var safeIDRe = regexp.MustCompile(`[^\w.-]`)

// Config is the minimal slice of AppConfig this package needs.
type Config struct {
	DebugLog bool
	LogDir   string
}

var (
	mu            sync.Mutex
	activeLogPath string
	warnedOnce    bool
)

// enabled reports whether logging is on for the given config.
func enabled(cfg Config) bool {
	return cfg.DebugLog && cfg.LogDir != ""
}

// warnOnce prints a single disabled-for-this-run notice to stderr. Caller holds mu.
func warnOnce(err error) {
	if warnedOnce {
		return
	}
	warnedOnce = true
	fmt.Fprintf(os.Stderr, "[debugLog] write failed (logging disabled for this run): %v\n", err)
}

// StartDebugLog primes per-session logging: prune old logs, choose the session
// file path, emit the session.start header, and return the path. No-op (returns
// "") when disabled. Call once per process after config load.
func StartDebugLog(cfg Config, header map[string]any) string {
	if !enabled(cfg) {
		return ""
	}
	mu.Lock()
	pruneOldLogs(cfg.LogDir)
	sessionID, _ := header["sessionId"].(string)
	if sessionID == "" {
		sessionID = randomID()
	}
	date := time.Now().UTC().Format("2006-01-02")
	safe := safeIDRe.ReplaceAllString(sessionID, "")
	if safe == "" {
		safe = "session"
	}
	activeLogPath = filepath.Join(cfg.LogDir, fmt.Sprintf("%s-%s.log", date, safe))
	path := activeLogPath
	mu.Unlock()

	LogDebug(cfg, "session.start", header)
	return path
}

// CurrentDebugLogPath returns the active log path once logging has started, or "".
func CurrentDebugLogPath() string {
	mu.Lock()
	defer mu.Unlock()
	return activeLogPath
}

// LogDebug appends one event line (plus any block values) to the session log.
// No-op unless logging is enabled. Catches all errors → warnOnce; never panics.
func LogDebug(cfg Config, event string, fields map[string]any) {
	if !enabled(cfg) {
		return
	}
	mu.Lock()
	defer mu.Unlock()

	if err := os.MkdirAll(cfg.LogDir, 0o700); err != nil {
		warnOnce(err)
		return
	}
	target := resolveTarget(cfg.LogDir)

	line := formatLine(event, fields)

	f, err := os.OpenFile(target, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o600)
	if err != nil {
		warnOnce(err)
		return
	}
	defer f.Close()
	if _, err := f.WriteString(line); err != nil {
		warnOnce(err)
	}
}

// resolveTarget reuses activeLogPath iff its dir matches logDir; else lazily opens
// a fresh <date>-<randomId>.log so stray writes still coalesce. Caller holds mu.
func resolveTarget(logDir string) string {
	if activeLogPath != "" && filepath.Dir(activeLogPath) == logDir {
		return activeLogPath
	}
	date := time.Now().UTC().Format("2006-01-02")
	activeLogPath = filepath.Join(logDir, fmt.Sprintf("%s-%s.log", date, randomID()))
	return activeLogPath
}

// formatLine renders one trace line:
//
//	<ISO-timestamp>  <event>[  k1=v1 ...]\n
//	[  <blockKey>:\n    <indented value>\n ...]
//
// Inline scalars: null/number/bool, or a string ≤120 chars without newline.
// Nil-valued fields are omitted (noise). Block values: objects/arrays via indented
// JSON, strings as-is.
//
// EVERY rendered string — inline and block alike — passes through redact.String before
// it is written, and block values are additionally capped at blockValueMax. Numbers and
// booleans skip redaction: they cannot carry a credential, and running a dozen regexes
// over "durationMs=38" on every line of a chatty trace is pure cost.
func formatLine(event string, fields map[string]any) string {
	ts := time.Now().UTC().Format("2006-01-02T15:04:05.000Z07:00")

	// Deterministic field order for stable output.
	keys := make([]string, 0, len(fields))
	for k := range fields {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	var inline strings.Builder
	type block struct {
		key, val string
	}
	var blocks []block

	for _, k := range keys {
		v := fields[k]
		if v == nil {
			// Distinguish a literal nil interface (omit) — but in Go an explicit
			// JSON null isn't representable as a nil here, so we render literal
			// "null" only for explicit Null sentinel; a nil value is omitted.
			continue
		}
		if scalar, ok := inlineScalar(v); ok {
			fmt.Fprintf(&inline, "  %s=%s", k, scalar)
			continue
		}
		// blockValue redacts as it renders (see there), so only the cap is applied here.
		// Redacting again would run 15 regexes a second time over the largest values in
		// the trace — precisely the payloads where that costs the most.
		blocks = append(blocks, block{key: k, val: redact.Cap(blockValue(v), blockValueMax)})
	}

	var b strings.Builder
	b.WriteString(ts)
	b.WriteString("  ")
	b.WriteString(event)
	b.WriteString(inline.String())
	b.WriteByte('\n')
	for _, bl := range blocks {
		fmt.Fprintf(&b, "  %s:\n", bl.key)
		for _, l := range strings.Split(bl.val, "\n") {
			fmt.Fprintf(&b, "    %s\n", l)
		}
	}
	return b.String()
}

// Null is a sentinel rendering as the literal "null" inline (distinct from an
// omitted nil field).
var Null = nullValue{}

type nullValue struct{}

// inlineScalar reports whether v renders inline, returning its rendering.
func inlineScalar(v any) (string, bool) {
	switch t := v.(type) {
	case nullValue:
		return "null", true
	case bool:
		return strconv.FormatBool(t), true
	case int:
		return strconv.Itoa(t), true
	case int64:
		return strconv.FormatInt(t, 10), true
	case float64:
		return strconv.FormatFloat(t, 'g', -1, 64), true
	case string:
		// Redact BEFORE the length test, not after: masking can change the length, and a
		// 130-char string that redacts to 110 should render inline like any other short
		// value rather than becoming an indented block for no visible reason.
		r := redact.String(t)
		if len(r) <= inlineMax && !strings.ContainsRune(r, '\n') {
			return r, true
		}
		return "", false
	default:
		return "", false
	}
}

// blockValue renders a non-inline value, REDACTED: a string through the free-text
// patterns, everything else walked structurally and re-marshaled as indented JSON.
//
// Structured values are walked rather than regexed for the same reason the audit path is:
// running the patterns over serialized JSON corrupted it (an env-assignment value ran to
// the next whitespace, and minified JSON has none). The caller applies the size cap.
func blockValue(v any) string {
	if s, ok := v.(string); ok {
		return redact.String(s)
	}
	data, err := json.MarshalIndent(redact.Value(v), "", "  ")
	if err != nil {
		// %v on an unmarshalable value can still print struct fields verbatim, so the
		// fallback gets the free-text pass rather than escaping redaction entirely.
		return redact.String(fmt.Sprintf("%v", v))
	}
	return string(data)
}

// pruneOldLogs deletes matching logs older than maxLogAgeMs by mtime. Best-effort;
// a missing dir is a no-op; never panics.
func pruneOldLogs(dir string) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return
	}
	cutoff := time.Now().UnixMilli() - maxLogAgeMs
	for _, e := range entries {
		if e.IsDir() || !sessionLogRe.MatchString(e.Name()) {
			continue
		}
		info, err := e.Info()
		if err != nil {
			continue
		}
		if info.ModTime().UnixMilli() < cutoff {
			_ = os.Remove(filepath.Join(dir, e.Name()))
		}
	}
}

// randomID returns an 8-char lowercase base36 id. The produced filename must
// satisfy sessionLogRe.
func randomID() string {
	const alphabet = "0123456789abcdefghijklmnopqrstuvwxyz"
	b := make([]byte, 8)
	for i := range b {
		b[i] = alphabet[rand.Intn(len(alphabet))]
	}
	return string(b)
}
